core/src/main/java/nu/yona/server/analysis/service/AnalysisEngineCacheService.java

```diff
@@ -4,30 +4,38 @@
  *******************************************************************************/
 package nu.yona.server.analysis.service;
 
-import java.util.Date;
-import java.util.List;
+import java.time.ZonedDateTime;
 import java.util.UUID;
 
 import org.springframework.cache.annotation.CachePut;
 import org.springframework.cache.annotation.Cacheable;
 import org.springframework.stereotype.Service;
 
-import nu.yona.server.analysis.entities.Activity;
+import nu.yona.server.analysis.entities.DayActivity;
+import nu.yona.server.analysis.entities.WeekActivity;
 
 @Service
 public class AnalysisEngineCacheService
 {
-	@Cacheable(value = "activities", key = "{#userAnonymizedID,#goalID}")
-	public Activity fetchLatestActivityForUser(UUID userAnonymizedID, UUID goalID, Date minEndTime)
+	@Cacheable(value = "dayActivities", key = "{#userAnonymizedID,#goalID,#zonedStartOfDay}")
+	public DayActivity fetchDayActivityForUser(UUID userAnonymizedID, UUID goalID, ZonedDateTime zonedStartOfDay)
 	{
-		List<Activity> results = Activity.getRepository().findLatestActivity(userAnonymizedID, goalID, minEndTime);
+		return DayActivity.getRepository().findOne(userAnonymizedID, goalID, zonedStartOfDay);
+	}
 
-		return results != null && !results.isEmpty() ? results.get(0) : null;
+	@CachePut(value = "dayActivities", key = "{#dayActivity.userAnonymizedID,#dayActivity.goalID,#dayActivity.zonedStartTime}")
+	public DayActivity updateDayActivityForUser(DayActivity dayActivity)
+	{
+		return DayActivity.getRepository().save(dayActivity);
+	}
+
+	public WeekActivity fetchWeekActivityForUser(UUID userAnonymizedID, UUID goalID, ZonedDateTime zonedStartOfWeek)
```

Comment:
Because the code looks so much alike :). The reason that it is not cached is that I'm not sure the activities of a whole week for all users fits in memory.